### ISSUE TITLE
"Unparsed triple" errors for models with reification and annotation properties

### ISSUE THREAD

**Issue by rpgoldman** [ontodev/robot#785]
I'm trying to use `merge` to merge together an ontology and a file with a set of entities that have annotation properties.  This second file is generated by flattening Turtle* to Turtle with `RDFstarTools`.

I'm getting floods of `INFO` messages about "Unparsed triple"s.

These seem to be for *any* triple whose predicate is `rdf:subject`, `rdf:predicate` or `rdf:object` or that are of the form *URL* `rdf:type` `rdf:Statement`.

*Should* I even expect this to work?

**Comment by rpgoldman** [comment]
P.S. Should I be using the Turtle* files directly, instead of flattening them first?

**Comment by jamesaoverton** [comment]
Thanks for the report. Can you provide some example triples and error messages, please?

**Comment by rpgoldman** [comment]
I will see what I can do -- the problems I had are on data that's not mine, so I have to do a little anonymization before I can post something.

**Comment by jamesaoverton** [comment]
Thanks. The larger issue is that RDF and OWL have two different ways to reify triples: 

https://stackoverflow.com/questions/45610092/owl-reification-vs-rdf-reification

Offhand I don't know how OWLAPI (which ROBOT is built on) handles RDF reification.

**Comment by rpgoldman** [comment]
There are a couple of threads on this in the archives of the OWLAPI developers' list:

https://sourceforge.net/p/owlapi/mailman/search/?q=reification

Unfortunately, the conversations are rambling, and I'm not sure if they ever reached conclusions, so I couldn't say what OWLAPI does and doesn't do based on this.

**Comment by rpgoldman** [comment]
OK, here's a minimum working example ttl file that cannot be processed by robot.  I stumbled on this using `robot merge` but for this MWE, you can just use `robot report` as a way of seeing that this file cannot be processed.  BTW, I put the turtle files in line because I couldn't find an "attach files" button.

I built this file by automatically translating a Turtle* file to this Turtle file.

```
@prefix :      <urn:robot:reificationExample:> .
@prefix owl:   <http://www.w3.org/2002/07/owl#> .
@prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix xsd:   <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .
:ModelEntity rdf:type owl:Class .
:ModelVariable rdf:type owl:Class .
:Bar rdf:type :ModelEntity .
:indicatedBy rdf:type owl:ObjectProperty ;
                                         :domain :ModelVariable ;
                                         :range :ModelEntity .
:specificProperty rdf:type owl:DatatypeProperty ;
                                              :rdfs:range xsd:string .
:propertyOne rdf:type owl:ObjectProperty .
:Foo rdf:type :ModelVariable ;
                                 :indicatedBy :Bar ;
                                 :label "Number of Other Vehicles (Causal Model Variable)" .
:Foo :indicatedBy :Bar .
_:Bb65616dcX2Dde2dX2D405aX2Da1aaX2Ddfbb25f03995 rdf:type rdf:Statement ;
                                        rdf:subject :Foo ;
                                        rdf:predicate :indicatedBy ;
                                        rdf:object :Bar ;
                                        :specificProperty ":propertyOne" .
```

The above file, robot doesn't like because of the notation used for the anonymous node for the reification (it seems to be misinterpreted as a HexBin block).  So we rename this to have a better name:

```
@prefix :      <urn:robot:reificationExample:> .
@prefix owl:   <http://www.w3.org/2002/07/owl#> .
@prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix xsd:   <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .
:ModelEntity rdf:type owl:Class .
:ModelVariable rdf:type owl:Class .
:Bar rdf:type :ModelEntity .
:indicatedBy rdf:type owl:ObjectProperty ;
                                         :domain :ModelVariable ;
                                         :range :ModelEntity .
:specificProperty rdf:type owl:DatatypeProperty ;
                                              :rdfs:range xsd:string .
:propertyOne rdf:type owl:ObjectProperty .
:Foo rdf:type :ModelVariable ;
                                 :indicatedBy :Bar ;
                                 :label "Number of Other Vehicles (Causal Model Variable)" .
:Foo :indicatedBy :Bar .
:Bb65616dcX2Dde2dX2D405aX2Da1aaX2Ddfbb25f03995 rdf:type rdf:Statement ;
                                        rdf:subject :Foo ;
                                        rdf:predicate :indicatedBy ;
                                        rdf:object :Bar ;
                                        :specificProperty ":propertyOne" .

```

*Now* robot doesn't like the file because it is missing `rdf:label` properties, so I add them:

```
@prefix :      <urn:robot:reificationExample:> .
@prefix owl:   <http://www.w3.org/2002/07/owl#> .
@prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix xsd:   <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .
:ModelEntity rdf:type owl:Class .
:ModelVariable rdf:type owl:Class .
:Bar rdf:type :ModelEntity .

:indicatedBy rdf:type owl:ObjectProperty ;
    rdfs:label "Property of reified statements." ;
    :domain :ModelVariable ;
    :range :ModelEntity .

:specificProperty rdf:type owl:DatatypeProperty ;
    rdfs:label "An example property." ;
    :rdfs:range xsd:string .

:propertyOne rdf:type owl:ObjectProperty .

:Foo rdf:type :ModelVariable ;
    :indicatedBy :Bar ;
    rdfs:label "Number of Other Vehicles (Causal Model Variable)" .

:Foo :indicatedBy :Bar .

:Bb65616dcX2Dde2dX2D405aX2Da1aaX2Ddfbb25f03995 rdf:type rdf:Statement ;
    rdfs:label "Reified statement." ;
    rdf:subject :Foo ;
    rdf:predicate :indicatedBy ;
    rdf:object :Bar ;
    :specificProperty ":propertyOne" .

```

and the final messages -- from `java -jar ../../tools/robot.jar -vv report --input mweprimeprime.ttl` -- we see the "Unparsed triple": 

```
$ java -jar ../../tools/robot.jar -vv report --input mweprimeprime.ttl
2020-12-04 17:14:09,135 INFO  org.semanticweb.owlapi.rdf.rdfxml.parser.OWLRDFConsumer - Unparsed triple: urn:robot:reificationExample:Bb65616dcX2Dde2dX2D405aX2Da1aaX2Ddfbb25f03995 -> http://www.w3.org/1999/02/22-rdf-syntax-ns#predicate -> urn:robot:reificationExample:indicatedBy
2020-12-04 17:14:09,135 INFO  org.semanticweb.owlapi.rdf.rdfxml.parser.OWLRDFConsumer - Unparsed triple: urn:robot:reificationExample:Bb65616dcX2Dde2dX2D405aX2Da1aaX2Ddfbb25f03995 -> http://www.w3.org/1999/02/22-rdf-syntax-ns#subject -> urn:robot:reificationExample:Foo
2020-12-04 17:14:09,135 INFO  org.semanticweb.owlapi.rdf.rdfxml.parser.OWLRDFConsumer - Unparsed triple: urn:robot:reificationExample:Bb65616dcX2Dde2dX2D405aX2Da1aaX2Ddfbb25f03995 -> http://www.w3.org/1999/02/22-rdf-syntax-ns#type -> http://www.w3.org/1999/02/22-rdf-syntax-ns#Statement
2020-12-04 17:14:09,135 INFO  org.semanticweb.owlapi.rdf.rdfxml.parser.OWLRDFConsumer - Unparsed triple: urn:robot:reificationExample:Bb65616dcX2Dde2dX2D405aX2Da1aaX2Ddfbb25f03995 -> http://www.w3.org/1999/02/22-rdf-syntax-ns#object -> urn:robot:reificationExample:Bar
2020-12-04 17:14:09,409 INFO  org.obolibrary.robot.QueryOperation - Named graph added: urn:uuid:70474929-c69e-409c-b8d8-040f4d72127d
Violations: 5
-----------------
ERROR:      3
WARN:       2
INFO:       0
Level	Rule Name	Subject	Property	Value
ERROR	missing_ontology_description	551a9ee6-6afc-430e-a8c4-5ec80d078677genid1	dc11:description
ERROR	missing_ontology_license	551a9ee6-6afc-430e-a8c4-5ec80d078677genid1	dc:license
ERROR	missing_ontology_title	551a9ee6-6afc-430e-a8c4-5ec80d078677genid1	dc11:title
WARN	missing_definition	urn:robot:reificationExample:indicatedBy	IAO:0000115
WARN	missing_definition	urn:robot:reificationExample:specificProperty	IAO:0000115

2020-12-04 17:14:09,646 ERROR org.obolibrary.robot.ReportCommand - Report failed!
```

**Comment by rpgoldman** [comment]
As a separate point, it is really bad that the parser can throw lines of input on the floor, and the user won't get a message about this unless verbosity is cranked up.

robot should scream at the user loud and clear if something is no good, instead of quietly ignoring it.

**Comment by jamesaoverton** [comment]
Thank you for the detailed report @rpgoldman. I have never used OWLAPI or ROBOT with RDF reification, only OWL reification, so this is news to me, and I spent some time trying to figure it out. Ultimately I think this is a problem with OWLAPI. I have a bit of a workaround at the end.

ROBOT builds on OWLAPI, and OWLAPI is doing all the work here. The "Unparsed triple" logging message is here:

https://github.com/owlcs/owlapi/blob/version4/parsers/src/main/java/org/semanticweb/owlapi/rdf/rdfxml/parser/OWLRDFConsumer.java#L1316

That's called in `dumpRemainingTriples()` and `endModel()`. OWLAPI just logs this as INFO, no exception, so ROBOT doesn't do anything special here. I agree with you that it's bad behaviour to drop triples with just an INFO-level log message.

I experimented with your examples. Running them through `robot convert`, the output is missing the `rdf:subject`, `rdf:predicate` and `rdf:object` triples, which is bad. Eventually I tried adding these declarations, and then those triples appeared in the output:

```
rdf:subject rdf:type owl:ObjectProperty .
rdf:predicate rdf:type owl:ObjectProperty .
rdf:object rdf:type owl:ObjectProperty .
```

Can you confirm that?

I've seen this sort of thing before when using OWLAPI: If predicates aren't given the types OWLAPI expects, and it can't infer the types, it sometimes ignores those predicates. My solution has always been to declare the types the OWLAPI wants. I found a few related issues: <https://github.com/owlcs/owlapi/search?q=%22rdf%3AProperty%22&type=issues>. I don't see what can be done in ROBOT to fix this -- I think it's an OWLAPI issue.

**Comment by rpgoldman** [comment]
I added your declarations block, and I still get the error:
```
@prefix :      <urn:logx:ontology:> .
@prefix qudt:  <http://qudt.org/schema/qudt/> .
@prefix stardog: <tag:stardog:api:> .
@prefix owl:   <http://www.w3.org/2002/07/owl#> .
@prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix xsd:   <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .


rdf:subject rdf:type owl:ObjectProperty .
rdf:predicate rdf:type owl:ObjectProperty .
rdf:object rdf:type owl:ObjectProperty .

... my turtle here...
```

E.g.,
```
2020-12-08 09:55:11,816 INFO  org.semanticweb.owlapi.rdf.rdfxml.parser.OWLRDFConsumer - Unparsed triple: urn:project:ontology:B2a29791cX2Db410X2D4804X2D9c94X2Db49db1b5afff -> http://www.w3.org/1999/02/22-rdf-syntax-ns#type -> http://www.w3.org/1999/02/22-rdf-syntax-ns#Statement
```

This `B2a29791cX2Db410X2D4804X2D9c94X2Db49db1b5afff` URI component is an example of a term that is generated by the translation of Turtle* to Turtle.

Here is a new version of the MWE:


```
@prefix :      <urn:robot:reificationExample:> .
@prefix owl:   <http://www.w3.org/2002/07/owl#> .
@prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix xsd:   <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .

rdf:subject rdf:type owl:ObjectProperty .
rdf:predicate rdf:type owl:ObjectProperty .
rdf:object rdf:type owl:ObjectProperty .

:ModelEntity rdf:type owl:Class .
:ModelVariable rdf:type owl:Class .
:Bar rdf:type :ModelEntity .

:indicatedBy rdf:type owl:ObjectProperty ;
    rdfs:label "Property of reified statements." ;
    :domain :ModelVariable ;
    :range :ModelEntity .

:specificProperty rdf:type owl:DatatypeProperty ;
    rdfs:label "An example property." ;
    :rdfs:range xsd:string .

:propertyOne rdf:type owl:ObjectProperty .

:Foo rdf:type :ModelVariable ;
    :indicatedBy :Bar ;
    rdfs:label "Number of Other Vehicles (Causal Model Variable)" .

:Foo :indicatedBy :Bar .

:Bb65616dcX2Dde2dX2D405aX2Da1aaX2Ddfbb25f03995 rdf:type rdf:Statement ;
    rdfs:label "Reified statement." ;
    rdf:subject :Foo ;
    rdf:predicate :indicatedBy ;
    rdf:object :Bar ;
    :specificProperty ":propertyOne" .
```

and here's the critical line of log output:

```
2020-12-08 10:00:21,434 INFO  org.semanticweb.owlapi.rdf.rdfxml.parser.OWLRDFConsumer - Unparsed triple: urn:robot:reificationExample:Bb65616dcX2Dde2dX2D405aX2Da1aaX2Ddfbb25f03995 -> http://www.w3.org/1999/02/22-rdf-syntax-ns#type -> http://www.w3.org/1999/02/22-rdf-syntax-ns#Statement
```

I tried adding:

```
rdf:Statement rdf:type owl:Class .
```

But still get the same error message.

Tried using `rdfs:Class` instead of `owl:Class`: same message.

**Comment by rpgoldman** [comment]
> I've seen this sort of thing before when using OWLAPI: If predicates aren't given the types OWLAPI expects, and it can't infer the types, it sometimes ignores those predicates. My solution has always been to declare the types the OWLAPI wants. I found a few related issues: [owlcs/owlapi/search?q=%22rdf%3AProperty%22&type=issues](https://github.com/owlcs/owlapi/search?q=%22rdf%3AProperty%22&type=issues). I don't see what can be done in ROBOT to fix this -- I think it's an OWLAPI issue.

I hesitate to suggest this, because it's so disgusting, but maybe robot needs to read the OWLAPI log and raise its own error condition if it sees any "Unparsed triple" messages.  Better, of course, would be for you to create an issue against OWLAPI and get them to at least give you the option of raising an exception when this happens.  I'm afraid doing that is way above my pay grade...

**Comment by rpgoldman** [comment]
Haha! I've got it!  OWLAPI can handle **OWL** reification, but not RDF reification (AFAICT).

I changed the reification block to:

```
:Bb65616dcX2Dde2dX2D405aX2Da1aaX2Ddfbb25f03995 a owl:Axiom ;
    rdfs:label "Reified statement." ;
    owl:annotatedSource :Foo ;
    owl:annotatedProperty :indicatedBy ;
    owl:annotatedTarget :Bar ;
    :specificProperty ":propertyOne" .
```

... and the error went away.

**Comment by rpgoldman** [comment]
Unfortunately, RDF* is defined in terms of *RDF* reification, not OWL reification, to the best of my knowledge. So now one needs a tool to translate RDF*/Turtle* to OWL reification.

"you are in a maze of twisty little passages all alike "

**Comment by jamesaoverton** [comment]
Yes, I use OWL reification all the time with OWLAPI and ROBOT, without any particular problems.

I don't know why there are two kinds of reification, seemingly equivalent. OWL doesn't support RDF datasets / quads, so while RDF* looks great, I don't expect direct OWL support.

I'll think some more about what can be done in ROBOT and/or OWLAPI about this.

**Comment by matentzn** [comment]
From a short [blogpost](https://douroucouli.wordpress.com/2020/09/11/edge-properties-part-1-reification/) around this issue by @cmungall 

> _Alternative Reification Vocabularies_. This includes the [OWL reification vocabulary](https://www.w3.org/TR/owl2-mapping-to-rdf/#Translation_of_Axioms_with_Annotations). It’s immensely depressing and confusing and under-appreciated that OWL did not adopt the RDF reification vocabulary, and the OWL stack fails horribly when we try and use the two together. Additionally OWL reification comes with annoying limitations (see [my answer on stack overflow](https://stackoverflow.com/questions/45610092/owl-reification-vs-rdf-reification) about RDF vs OWL reification).

OWL API is not for working with RDF. Its just for working with OWL/OBO, and as long as ROBOT is an OWL/OBO tool, I don't think we can cover RDF. I feel your pain for a lack of tool support that bridges the two worlds more.

**Comment by rpgoldman** [comment]
The problem for me is that I need to translate Turtle* into Turtle (don't I?) to use normal RDF tools and the only translator I have found uses RDF reification.

I can write a simple string replacer that turns RDF reification into owl reification but ... 🤮

I *do* think that the "I'm quietly throwing your precious triples into the bitbucket" issue needs to be fixed, even if it requires some more complex interaction with OWLAPI's parser.

For that matter, throwing an error when anyone tries to use RDF reification would be helpful, too, since one could have a very informative error message that would point the poor user towards an easy fix.

**Comment by balhoff** [comment]
> I do think that the "I'm quietly throwing your precious triples into the bitbucket" issue needs to be fixed, even if it requires some more complex interaction with OWLAPI's parser.

I opened an issue: https://github.com/owlcs/owlapi/issues/988

I also think we should add some sort of discussion to the docs about the dangers of trying to use ROBOT as a tool for arbitrary RDF.

**Comment by balhoff** [comment]
@rpgoldman for merging RDF you may find Jena's [`rdfcat`](https://www.javadoc.io/static/org.apache.jena/jena-core/2.11.0/index.html?jena/rdfcat.html) tool useful.

**Comment by jamesaoverton** [comment]
@rpgoldman Would #787 address your needs? If not, please let us know how it falls short. If it does cover your needs, I'd prefer to close this issue in favour of that one.

**Comment by rpgoldman** [comment]
> @rpgoldman Would #787 address your needs? If not, please let us know how it falls short. If it does cover your needs, I'd prefer to close this issue in favour of that one.

Absolutely: that would be great.

It would be nice if you could add a special check for the use of RDF reification and a targeted error message. It's really hard for someone to know from outside that OWL reification is required and RDF reification is not supported. And once one *does* know, it's really easy (string replace!) to fix in one's ontology.

**Comment by jamesaoverton** [comment]
Yes. I want ROBOT to inspect the unparsed triples and try to give a useful message based on what it sees. RDF reification would be one such case. I'll add a note about this to the other issue.

**Comment by cmungall** [comment]
@rpgoldman I share your frustration re difference between owl and rdf reification. We are increasingly using rdf* which necessitates two stacks and interconversion.

Simple IRI replacement may not suffice. For example, if you give your RDF statements IRIs, these cannot be preserved directly in OWL reification (without creating a new triple, e.g. `?axiom schema:iri ?iri .`). With OWL reification, axioms must be blank nodes. I have no idea why. Details like this are obscured in a complex web of W3 specs and I generally find out about them through unpleasant empirical discovery.

**Comment by rpgoldman** [comment]
@cmungall Thanks for the pointer.  This is interesting. The tool I used to translate the Turtle* to Turtle generated what I think were blank nodes (their IDs were of the form `_:xxxx`, if I recall correctly).  But OWLAPI's parser didn't like this syntax for blank nodes, and I replaced them with "real" IDs (by throwing away the underscores).  Interestingly, that did *not* break OWLAPI, once I rewrote rdf:subject, predicate and statement to their OWL reification equivalents.  But from what you say above, it should have been an error.
